plugin/src/software/aws/toolkits/eclipse/amazonq/chat/ChatCommunicationManager.java

```diff
@@ -233,21 +231,22 @@ private void sendErrorToUi(final String tabId, final Throwable exception) {
     }
 
     public void setChatUiRequestListener(final ChatUiRequestListener listener) {
-        chatUiRequestListener = listener;
+        chatUiRequestListenerFuture.complete(listener);
     }
 
     public void removeListener() {
-        chatUiRequestListener = null;
+        chatUiRequestListenerFuture = new CompletableFuture<>();
     }
 
     /*
      * Sends message to Chat UI to show in webview
      */
     public void sendMessageToChatUI(final ChatUIInboundCommand command) {
-        if (chatUiRequestListener != null) {
-            String message = jsonHandler.serialize(command);
-            chatUiRequestListener.onSendToChatUi(message);
-        }
+        String message = jsonHandler.serialize(command);
+        chatUiRequestListenerFuture.thenApply(listener -> {
```

Comment:
this approach seems very odd IMO. Traditionally listeners in an event subscriber pattern are not intended to be completable futures. In this case, the webview class is the listener to an event. Why do we need to return the listener on L248?

Reply:
on second thought, I think the fix is simply moving the [removeListener call here ](https://github.com/aws/amazon-q-eclipse/blob/55bba2ab47b22b924a342631d8c1d8d6747e87d9/plugin/src/software/aws/toolkits/eclipse/amazonq/views/AmazonQChatWebview.java#L279)to be within the if (canDisposeState) check .
The reason is when the view is closed the listener is unhooked which means messages sent in the time between it is closed and opened can get dropped leading to the view getting stuck in an in-progress state.
Can you try the above suggestion and validate it works as expected?

Reply:
Let's address this as part of the larger refactor to the views @taldekar.